src/mlpack/tests/main_tests/nbc_test.cpp

```diff
@@ -0,0 +1,216 @@
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Row<size_t>>("output").n_rows, 1);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output_probs").n_rows, 2);
+}
+
+/**
+ * Ensure that saved model can be used again.
+ */
+BOOST_AUTO_TEST_CASE(NBCModelReuseTest)
+{
+  arma::mat inputData;
+  if (!data::Load("trainSet.csv", inputData))
+    BOOST_FAIL("Cannot load train dataset trainSet.csv!");
+
+  arma::mat testData;
+  if (!data::Load("testSet.csv", testData))
+    BOOST_FAIL("Cannot load test dataset testSet.csv!");
+
+  // Delete the last row containing labels from test dataset.
+  testData.shed_row(testData.n_rows - 1);
+
+  size_t testSize = testData.n_cols;
+
+  // Input training data.
+  SetInputParam("training", std::move(inputData));
+
+  // Input test data.
+  SetInputParam("test", std::move(testData));
+
+  mlpackMain();
+
+  arma::Row<size_t> output;
+  arma::mat output_probs;
+  output = std::move(CLI::GetParam<arma::Row<size_t>>("output"));
+  output_probs = std::move(CLI::GetParam<arma::mat>("output_probs"));
+
+  // Reset passed parameters.
+  CLI::GetSingleton().Parameters()["training"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["test"].wasPassed = false;
+
+  if (!data::Load("testSet.csv", testData))
+    BOOST_FAIL("Cannot load test dataset testSet.csv!");
+
+  // Delete the last row containing labels from test dataset.
+  testData.shed_row(testData.n_rows - 1);
+
+  // Input trained model.
+  SetInputParam("test", std::move(testData));
+  SetInputParam("input_model",
+                std::move(CLI::GetParam<NBCModel>("output_model")));
+
+  mlpackMain();
+
+  // Check that number of output points are equal to number of input points.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Row<size_t>>("output").n_cols,
+                      testSize);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output_probs").n_cols,
+                      testSize);
+
+  // Check output have only single row.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Row<size_t>>("output").n_rows, 1);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output_probs").n_rows, 2);
+
+  // Check that initial output and final output
+  // matrix using saved model are same.
+  CheckMatrices(output, CLI::GetParam<arma::Row<size_t>>("output"));
+  CheckMatrices(output_probs, CLI::GetParam<arma::mat>("output_probs"));
+}
+
+/**
+ * Make sure only one of training data or pre-trained model is passed.
+ */
+BOOST_AUTO_TEST_CASE(NBCTrainingVerTest)
+{
+  arma::mat inputData;
+  if (!data::Load("trainSet.csv", inputData))
+    BOOST_FAIL("Cannot load train dataset trainSet.csv!");
+
+  // Input training data.
+  SetInputParam("training", std::move(inputData));
+
+  mlpackMain();
+
+  // Input pre-trained model.
+  SetInputParam("input_model",
+                std::move(CLI::GetParam<NBCModel>("output_model")));
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
```

Comment:
We should also add a test for the incremental variance option.  That one can be simple: you can just train with and without incremental variance and ensure the output predictions and probabilities are the same.

Reply:
Done!